url8012c8d031_dmbb966_BBMS_tgz-pJ8-testbed_JNEATtestJ8/src/jneat/Population.java

```diff
@@ -0,0 +1,767 @@
+		
+		return buf.toString();
+	}
+	
+	public void speciate() {
+		int counter = 0;	// For number of species
+		Species newSpecies = null;
+		
+		Iterator<Organism> itr_organism = organisms.iterator();
+		while (itr_organism.hasNext()) {
+			Organism _organism = itr_organism.next();
+			// If the list species is empty, create the first
+			if (species.isEmpty()) {
+				newSpecies = new Species(++counter);				
+				newSpecies.organisms.add(_organism);
+				
+				species.add(newSpecies);
+				_organism.species = newSpecies;		
+			}
+			else {
+				Iterator<Species> itr_species = species.iterator();
+				boolean done = false;
+				
+				while (!done && itr_species.hasNext()) {
+					Species _species = itr_species.next();
+					Organism compare_org = _species.organisms.firstElement();
+					double curr_compat = _organism.genome.Compatibility(compare_org.genome);
+					
+					if (curr_compat < JNEATGlobal.p_compat_threshold) {
+						// Found compatible species - add this organism
+						_species.organisms.add(_organism);
+						_organism.species = _species;
+						done = true;
+					}
+				}
+				
+				// If no compatible species found, create a new one
+				if (!done) {
+					newSpecies = new Species(++counter);
+					species.add(newSpecies);
+					newSpecies.organisms.add(_organism);
+					_organism.species = newSpecies;
+				}
+			}
+		}
+		
+		last_species = counter; // Keep track of the highest species
+	}
+	
+	/** Adds a new species to this population along with the first organism in this new species*/
+	public void AddSpecies(Species newSpecies, Organism baby) {
+		last_species++;
+		species.add(newSpecies);
+		newSpecies.organisms.add(baby);
+		baby.species = newSpecies;
+	}
+	
+	public void Verify() {
+		Iterator<Organism> itr_organism = organisms.iterator();
+		while (itr_organism.hasNext()) {
+			itr_organism.next().genome.verify();
+		}
+	}
+	
+	/** Returns the current node ID and increments it.*/
+	public int getCurNodeID_Inc() {
+		return cur_node_id++;
+	}
+	
+	
+	/** Returns the current innovation number and increments it.*/
+	public int getCurInnov_Inc() {
+		return cur_innov_num++;
+	}
+	
+	public String PrintPopulation() {
+		String ret = "";
+		
+		ret += "\n\n\n\t\t >>>POPULATION<<<";
+		ret += "\n\n\t This population has " + organisms.size() + " organisms, ";
+		ret += species.size() + " species, with a population size: of " + population_size + "\n";
+		
+		Iterator<Organism> itr_organism = organisms.iterator();
+		while (itr_organism.hasNext()) {
+			ret += itr_organism.next().PrintOrganism() + "\n";
+		}
+		
+		Iterator<Species> itr_species = species.iterator();
+		while (itr_species.hasNext()) {
+			ret += itr_species.next().PrintSpecies() + "\n";
+		}
+		
+		return ret;
+	}
+	
+	
+	
+	/* Loads a population from a file at the designated path
+	 */
+	public Population(Path p) {
+		Charset cSet = Charset.forName("US-ASCII");
+		
+		try {
+			BufferedReader reader = Files.newBufferedReader(p,  cSet);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *reader*. Currently, there are execution paths that do not contain closure statements, for example, when an else condition at lines 600, 601 evaluate to true. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).